Moller-Plesset/DF-MP2_NAF.py

```diff
@@ -0,0 +1,210 @@
+__license__ = "BSD-3-Clause"
+__date__ = "2018-11-29"
+
+import time
+import numpy as np
+np.set_printoptions(precision=5, linewidth=200, suppress=True)
+import psi4
+
+# Set memory & output
+psi4.set_memory('2 GB')
+psi4.core.set_output_file('output.dat', False)
+
+mol = psi4.geometry(""" 
+C    1.39410    0.00000   0.00000
+C    0.69705   -1.20732   0.00000
+C   -0.69705   -1.20732   0.00000
+C   -1.39410    0.00000   0.00000
+C   -0.69705    1.20732   0.00000
+C    0.69705    1.20732   0.00000
+H    2.47618    0.00000   0.00000
+H    1.23809   -2.14444   0.00000
+H   -1.23809   -2.14444   0.00000
+H   -2.47618    0.00000   0.00000
+H   -1.23809    2.14444   0.00000
+H    1.23809    2.14444   0.00000
+symmetry c1
+""")
+
+# Basis used in mp2 density fitting
+psi4.set_options({'basis': 'aug-cc-pVDZ', 'df_basis_mp2': 'aug-cc-pvdz-ri'})
+
+check_energy = True
+
+print('\nStarting RHF...')
+t = time.time()
+RHF_E, wfn = psi4.energy('SCF', return_wfn=True)
+print('...RHF finished in %.3f seconds:   %16.10f' % (time.time() - t, RHF_E))
+
+# Grab data from Wavfunction clas
+ndocc = wfn.nalpha()
+orbital_basis = wfn.basisset()
+nbf = wfn.nso()
+nvirt = nbf - ndocc
+
+# Split eigenvectors and eigenvalues into o and v
+eps_occ = np.asarray(wfn.epsilon_a_subset("AO", "ACTIVE_OCC"))
+eps_vir = np.asarray(wfn.epsilon_a_subset("AO", "ACTIVE_VIR"))
+
+# Build DF tensors
+print('\nBuilding DF ERI tensor Qov...')
+t = time.time()
+C = wfn.Ca()
+
+# Build instance of MintsHelper
+mints = psi4.core.MintsHelper(orbital_basis)
+zero_bas = psi4.core.BasisSet.zero_ao_basis_set()
+
+# build auxiliary basis set object
+aux_basis = psi4.core.BasisSet.build(
+    mol, "DF_BASIS_MP2", "", "RIFIT",
+    psi4.core.get_global_option('df_basis_mp2'))
+naux = aux_basis.nbf()
+
+# Build (P|pq) raw 3-index ERIs, dimension (1, Naux, nbf, nbf)
+Ppq = mints.ao_eri(zero_bas, aux_basis, orbital_basis, orbital_basis)
+
+# Build Coulomb metric but only invert, dimension (1, Naux, 1, Naux)
+metric = mints.ao_eri(zero_bas, aux_basis, zero_bas, aux_basis)
+metric.power(-1.0, 1.e-14)
+
+# Remove excess dimensions of Ppq, & metric
+Ppq = np.squeeze(Ppq)
+metric = np.squeeze(metric)
+
+# paper uses transpose of Ppq, so we adapt for now
+Ppq = np.reshape(Ppq, (naux, nbf * nbf)).T
+print("I  = (pq|P) dim:", Ppq.shape)
+
+# cholesky decomp of inverse metric
+L = np.linalg.cholesky(metric)
+print("L  = cholesky[(P|Q)^1 ]dim:", L.shape)
+
+# Form intermediate W'= I^t*I
+Wp = np.dot(Ppq.T, Ppq)
+print("W' = (P|P) dim:", Wp.shape)
+
+# form W proper
+W = np.dot(np.dot(L.T, Wp), L)
+print("W  = (Q|Q) dim:", W.shape)
+
+# from N(bar) from eigenvectors of W
+# epsilon threshold is supposed to be in the range of 10^-2 to 10^-4
+e_val, e_vec = np.linalg.eigh(W)
+eps = 1e-2
+print('epsilon = %.3e ' % (eps))
+nskipped = 0
+Ntmp = np.zeros((naux, naux))
+naux2 = 0
+for n in range(naux):
+    if (abs(e_val[n]) > eps):
+        Ntmp[:, naux2] = e_vec[:, n]
+        naux2 += 1
```

Comment:
This might be better written as:

```
mask = e_val[e_val > 1.e-2]
naux2 = mask.sum()
Nbar = e_vec[:, mask]
```

Reply:
here mask holds the actual values, not the indices. I googled `np.argwhere` but need to figure out how it works correctly another time. quick test showed `Nbar` gets one extra dimension.

Reply:
Mask holds a boolean array which can index another NumPy array:

```python
>>> a
array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])
>>> a[:, np.array([True, False, True, False])]
array([[ 0,  2],
       [ 4,  6],
       [ 8, 10],
       [12, 14]])
```

I might be mistaken on the exact operation you are trying to accomplish above, but best to apply it through broadcasting.

Reply:
Yes, I want an array that I can use to index, and `np.argwhere` does the trick now. The suggested syntax returns the actual values.

```
>>> print(v)
[1.  0.5 0.3]
>>> v[v < 1 ]
array([0.5, 0.3])
>>> np.argwhere(v < 1 )
array([[1],
       [2]])
```